kingpin/actors/rightscale/alerts.py

```diff
@@ -370,3 +372,202 @@ def _execute(self):
+            'vote_tag': {'type': 'string'},
+            'vote_type': {'enum': ['grow', 'shrink']}
+        }
+    }
+
+
+class AlertSpecBase(base.EnsurableRightScaleBaseActor):
+
+    """Extremely simple AlertSpec creation actor.
+
+    This actor isn't really meant to be instantiated on its own -- it provides
+    the base functionality though for creating, deleting and updating an
+    AlertSpec on a given RightScale resource. The resource can be either a
+    Server Array, Server Template, Instance or Deployment.
+
+    **Options**
+
+    :href:
+      The RightScale HREF for the resource you wish to apply the Alert Spec to.
+
+    :state:
+      (str) Either `present` or `absent`
+
+    :spec:
+      A dictionary that conforms to the
+      :py:mod:`~kingpin.actors.rightscale.alerts.AlertSpecSchema`.
+
+    **Examples**
+
+    .. code-block:: json
+
+       { "actor": "rightscale.alerts.AlertSpecBase",
+         "options": {
+           "href": "/api/server_arrays/abcd1234",
+           "spec": {
+                "name": "Instance Stranded",
+                "description": "Alert if an instance enders a stranded",
+                "file": "RS/server-failure",
+                "variable": "state",
+                "condition": "==",
+                "threshold": "stranded",
+                "duration": 2,
+                "escalation_name": "critical"
+           }
+         }
+       }
+
+    """
+
+    all_options = {
+        'href': (str, None, 'RightScale Resource HREF to act on'),
+        'spec': (AlertSpecSchema, None, 'The actual Alert Spec definition itself.')
+    }
+    unmanaged_options = ['href']
+    desc = "AlertSpec: {spec[name]}"
+
+    def __init__(self, *args, **kwargs):
+        super(AlertSpecBase, self).__init__(*args, **kwargs)
+        self.changed = False
+
+    @gen.coroutine
+    def _precache(self):
+        name = self.option('spec').get('name')
+        href = self.option('href')
+
+        # Generate a fully populated set of parameters by including the href
+        # that was supplied to us.
+        desired_spec = dict(self.option('spec'))
+        desired_spec['subject_href'] = href
+        self.desired_params = self._generate_rightscale_params(
+            'alert_spec', desired_spec)
+
+        # Search for the existing spec
+        log.debug('Searching for AlertSpec matching: %s' % name)
+        self.spec = yield self._client.find_by_name_and_keys(
+            self._client._client.alert_specs,
+            name=name, exact=True, subject_href=href)
+
+        if not self.spec:
+            log.debug('AlertSpec matching "%s" could not be found.' % name)
+            self.spec = None
+        else:
+            log.debug('Got AlertSpec: %s' % self.spec)
+
+    @gen.coroutine
+    def _get_state(self):
+        if self.spec:
+            raise gen.Return('present')
+        raise gen.Return('absent')
+
+    @gen.coroutine
+    def _set_state(self):
+        if self.option('state') == 'absent':
```

Comment:
Renamed `self.spec` to `self.existing_spec` to help clarify how `_get_state()` works.